### PR TITLE
SIP authentication with token in call message

### PR DESCRIPTION
**Pull request by PavelAtCarbyne**
sip plugin will test siptoken if token_auth_secret is defined in janus.plugin.sip.jcfg

### CODE REVIEW

File: html/siptest.js [hunk 1]
```diff
@@ -650,7 +662,14 @@ function doCall() {
 				//				"AnotherHeader": "another string"
 				//			}
 				//		};
-				var body = { request: "call", uri: $('#peer').val() };
+                const token = getJanusSipToken('P-Asserted-Voicenter-Callerid,0730000000', $('#peer').val());
+                var body = { request: "call", uri: $('#peer').val(),
+                             siptoken: token,
+                             headers: {
+                             "P-Asserted-Voicenter-Callerid": "0730000000",
```
Comment (shmuel7) [inline html/siptest.js]: use the real parameters

File: plugins/janus_sip.c [hunk 2]
```diff
@@ -1936,6 +2002,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+        JANUS_LOG(LOG_INFO, "janus_sip: auth : fail, one of params is NULL\n");
+        return FALSE;
+    }
+    
+    gchar **parts = g_strsplit(token, ":", 2);
+    gchar **data = NULL;
+    /* Token should have exactly one data and one hash part */
+    if(!parts[0] || !parts[1] || parts[2]) {
+        JANUS_LOG(LOG_INFO, "janus_sip: auth: fail, Token should have exactly one data and one hash part \n");
+        goto fail;
```
Comment (shmuel7) [inline plugins/janus_sip.c]: CLEANUP macro instead of log+goto

File: plugins/janus_sip.c [hunk 3]
```diff
@@ -1272,7 +1332,13 @@ int janus_sip_init(janus_callbacks *callback, const char *config_path) {
 		janus_config_print(config);
 
 		janus_config_category *config_general = janus_config_get_create(config, NULL, janus_config_type_category, "general");
-		janus_config_item *item = janus_config_get(config, config_general, janus_config_type_item, "local_ip");
+                janus_config_item *item = janus_config_get(config, config_general, janus_config_type_item, "token_auth_secret");
+                if(item && item->value) {
+                    auth_enabled = TRUE;
+                    auth_secret = g_strdup(item->value);
```
Comment (shmuel7) [inline plugins/janus_sip.c]: Free string

File: html/siptest.js (hunk 1 above)
Comment (PavelAtCarbyne) [reply html/siptest.js]: P-Asserted-Identity 
123456789

File: html/siptest.js [hunk 4]
```diff
@@ -650,7 +662,14 @@ function doCall() {
 				//				"AnotherHeader": "another string"
 				//			}
 				//		};
-				var body = { request: "call", uri: $('#peer').val() };
+                const token = getJanusSipToken('P-Asserted-Voicenter-Callerid,0730000000', $('#peer').val());
```
[the remark this answers is not in the document]
Comment (PavelAtCarbyne) [reply html/siptest.js]: P-Asserted-Identity 
123456789

File: plugins/janus_sip.c [hunk 5]
```diff
@@ -2376,7 +2498,18 @@ static void *janus_sip_handler(void *data) {
 			/* Check if the INVITE needs to be enriched with custom headers */
 			char custom_headers[2048];
 			janus_sip_parse_custom_headers(root, (char *)&custom_headers);
-			/* SDES-SRTP is disabled by default, let's see if we need to enable it */
+            
+            /* Sip token  validation */
```
Comment (PavelAtCarbyne) [inline plugins/janus_sip.c]: spacing

File: plugins/janus_sip.c [hunk 6]
```diff
@@ -2376,7 +2498,18 @@ static void *janus_sip_handler(void *data) {
...
+            json_t *sip_token = json_object_get(root, "siptoken");
+            const char *sip_token_text = json_string_value(sip_token);
+            char headers_text[2048];
```
Comment (PavelAtCarbyne) [inline plugins/janus_sip.c]: initialize with zeros by = { 0 }

File: plugins/janus_sip.c [hunk 7]
```diff
@@ -1110,6 +1116,60 @@ static void janus_sip_random_string(int length, char *buffer) {
 	}
 }
 
+static  void janus_sip_get_all_key_values_from_custom_headers(json_t *root,  char *custom_value_out) {
+    custom_value_out[0] = '\0';
```
Comment (PavelAtCarbyne) [inline plugins/janus_sip.c]: remove this - you zeroed before

File: plugins/janus_sip.c [hunk 8]
```diff
@@ -1110,6 +1116,60 @@ static void janus_sip_random_string(int length, char *buffer) {
...
+            void *iter = json_object_iter(headers);
+            while(iter != NULL) {
+                key = json_object_iter_key(iter);
+                value = json_object_get(headers, key);
+                if(value == NULL || !json_is_string(value)) {
+                    JANUS_LOG(LOG_WARN, "Skipping header '%s': value is not a string\n", key);
+                    iter = json_object_iter_next(headers, iter);
+                    continue;
+                }
+                char h[255];
```
Comment (PavelAtCarbyne) [inline plugins/janus_sip.c]: zero array

Comment (PavelAtCarbyne) [reply plugins/janus_sip.c]: 255 and 2048 define consts

Comment (PavelAtCarbyne) [reply plugins/janus_sip.c]: spacing in func

File: plugins/janus_sip.c [hunk 9]
```diff
@@ -1133,37 +1134,9 @@ static  void janus_sip_get_all_key_values_from_custom_headers(json_t *root,  cha
...
-                    iter = json_object_iter_next(headers, iter);
-                    continue;
-                }
-                if(!strcmp(key, custom_key_in)) {
-                    char h[255];
-                    g_snprintf(h, 255, "%s", json_string_value(value));
-                    g_strlcat(custom_value_out, h, 2048);
-                    return;
-                }
+                char h[JANUS_SIP_SMALL_BUFFER_SIZE];
```
Comment (shmuel7) [inline plugins/janus_sip.c]: change to header

File: plugins/janus_sip.c [hunk 10]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
 	g_atomic_int_set(&session->hangingup, 0);
 }
 
+static gboolean janus_auth_check_signature(const char *token, const char *headers, const char *uri) {
+    if (!auth_enabled || auth_secret == NULL) {
+        JANUS_LOG(LOG_INFO, "janus_sip: auth : TRUE ,autentication not enabled \n");
+        return TRUE;
+    }
+    
+    if (token ==NULL || headers==NULL || uri==NULL ) {
```
Comment (shmuel7) [inline plugins/janus_sip.c]: spacing

File: plugins/janus_sip.c [hunk 11]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+    /* Verify timestamp */
+    gint64 expiry_time = atoi(data[0]);
+    gint64 real_time = janus_get_real_time() / 1000000;
+    if(real_time > expiry_time) {
+        JANUS_LOG(LOG_INFO, "janus_sip: auth: fail,  Verify timestamp\n");
+        goto fail;
+    }
+    
+    /* prepare message for compare with signature : <timestamp>,<src>,<dst> */
+    char message[4096];
```
Comment (shmuel7) [inline plugins/janus_sip.c]: const and init

File: plugins/janus_sip.c [hunk 12]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+    
+    if (token ==NULL || headers==NULL || uri==NULL ) {
+        JANUS_LOG(LOG_INFO, "janus_sip: auth : fail, one of params is NULL\n");
+        return FALSE;
+    }
+    
+    gchar **parts = g_strsplit(token, ":", 2);
+    gchar **data = NULL;
+    /* Token should have exactly one data and one hash part */
+    if(!parts[0] || !parts[1] || parts[2]) {
```
Comment (shmuel7) [inline plugins/janus_sip.c]: parts[2] - remove this check it could cause sigsev

File: plugins/janus_sip.c [hunk 13]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+        JANUS_LOG(LOG_INFO, "janus_sip: auth: fail, Token should have exactly one data and one hash part \n");
+        goto fail;
+    }
+    data = g_strsplit(parts[0], ",", 1);
```
Comment (shmuel7) [inline plugins/janus_sip.c]: delete this code it does nothing. you can use parts[0] instead

File: plugins/janus_sip.c [hunk 14]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+    gint64 real_time = janus_get_real_time() / 1000000;
+    if(real_time > expiry_time) {
+        JANUS_LOG(LOG_INFO, "janus_sip: auth: fail,  Verify timestamp\n");
+        goto fail;
+    }
+    
+    /* prepare message for compare with signature : <timestamp>,<src>,<dst> */
+    char message[4096];
+    message[0] = '\0';
+    g_snprintf(message, 4096, "%s,%s%s", data[0],headers,uri);
```
Comment (shmuel7) [inline plugins/janus_sip.c]: use const

File: plugins/janus_sip.c [hunk 15]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+    JANUS_LOG(LOG_INFO, "janus_sip: auth: left side of token:%s \n", message);
+    /* Verify HMAC-SHA256 */
+    unsigned char signature[EVP_MAX_MD_SIZE];
+    
+    unsigned int len;
```
Comment (shmuel7) [inline plugins/janus_sip.c]: intializae with = 0

File: plugins/janus_sip.c [hunk 16]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+        goto fail;
+    }
+    
+    /* prepare message for compare with signature : <timestamp>,<src>,<dst> */
+    char message[4096];
+    message[0] = '\0';
+    g_snprintf(message, 4096, "%s,%s%s", data[0],headers,uri);
+    JANUS_LOG(LOG_INFO, "janus_sip: auth: left side of token:%s \n", message);
+    /* Verify HMAC-SHA256 */
+    unsigned char signature[EVP_MAX_MD_SIZE];
```
Comment (shmuel7) [inline plugins/janus_sip.c]: initialize = { 0 }

File: plugins/janus_sip.c [hunk 17]
```diff
@@ -1936,6 +1975,62 @@ static void janus_sip_hangup_media_internal(janus_plugin_session *handle) {
...
+    
+    unsigned int len;
+    HMAC(EVP_sha256(), auth_secret, strlen(auth_secret), (const unsigned char*)message, strlen(message), signature, &len);
+    gchar *base64 = g_base64_encode(signature, len);
```
Comment (shmuel7) [inline plugins/janus_sip.c]: check if not null on all output params

File: plugins/janus_sip.c [hunk 18]
```diff
@@ -2376,7 +2471,18 @@ static void *janus_sip_handler(void *data) {
 			/* Check if the INVITE needs to be enriched with custom headers */
 			char custom_headers[2048];
 			janus_sip_parse_custom_headers(root, (char *)&custom_headers);
-			/* SDES-SRTP is disabled by default, let's see if we need to enable it */
+                        /* Sip token  validation */
+                        json_t *sip_token = json_object_get(root, "siptoken");
+                        const char *sip_token_text = json_string_value(sip_token);
+                        char headers_text[JANUS_SIP_LARGE_BUFFER_SIZE];
```
Comment (shmuel7) [inline plugins/janus_sip.c]: = { 0 }

**Pull request merged**